[Woo POS] Cache `ItemList` thumbnails. Retrieve from cache for `CartView`

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13915
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
This PR addresses the design feedback on https://github.com/woocommerce/woocommerce-ios/issues/13859 regarding cart images loading with a delay compared to item list images.

In order to resolve this, we rely on Kingfisher's built-in caching mechanism to store the Item List thumbnails temporarily, then retrieve them if needed for the Cart.

https://github.com/user-attachments/assets/c45417e3-566b-41c5-8269-959966408005

## Changes
- Added `CartItemRowViewModel`. This is a wrapper around the existing `CartItem`, and loads the appropriate image if can be located in cache.
- Updated the `ItemRowView` to render a cached image if this exists.
- Usage of KingFisher's cache via `ImageCache.default` as caching system, and `DefaultImageProcessor` to apply no special processing for cached images. The use of a custom processor also requires to update the caching setup, by using the default processor we can use the default image cache straight-away.

I have not added new unit tests. The only addition that we could test would be the `CartItemRowViewModel`, but the only methods that are exposed right now would be testing the internals of Kingfisher's caching.

## Testing information
- Load POS
- Tap on products with images that are loaded in the Item List
- Observe that the images for the products added to the cart show with no delay.
- For easier visuals, you can also add a hint like `.border(.red, width: 2.0)` to the Images loaded from cache, [here](https://github.com/woocommerce/woocommerce-ios/blob/e4b918b00c7728b951a31941a7489f63b9abb228/WooCommerce/Classes/POS/Presentation/ItemRowView.swift#L71)

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Caches ItemList thumbnails, then retrieves from cache on CartView

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Use cached images if found

The usage of `ResizingImageProcessor` seems to break the built-in Kingfisher caching, so we add the option to use `DefaultImageProcessor` conditionally when we need to.

iamgabrielma committed:
Use default processor ID

iamgabrielma committed:
Extract CartItemRowViewModel to own file

iamgabrielma committed:
Remove duplication on ItemRowView init

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/ItemRowView.swift:
```diff
@@ -59,17 +61,24 @@ struct ItemRowView: View {
-                   height: Constants.productCardSize * scale)
-            .clipped()
+        } else if let imageSource = viewModel.cartItem.item.productImageSource {
+            if let hasImageInCache = viewModel.cachedImage {
+                Image(uiImage: hasImageInCache)
+                    .resizable()
+                    .scaledToFill()
+                    .frame(width: min(Constants.productCardSize * scale, Constants.maximumProductCardSize),
+                           height: Constants.productCardSize * scale)
+                    .clipped()
+            } else {
+                ProductImageThumbnail(productImageURL: URL(string: imageSource),
+                                      productImageSize: Constants.productCardSize,
+                                      scale: scale,
+                                      foregroundColor: .clear)
+                .frame(width: min(Constants.productCardSize * scale, Constants.maximumProductCardSize),
+                       height: Constants.productCardSize * scale)
+                .clipped()
```
There's a bit of duplication here. I though of moving the logic to `ProductImageThumbnail` but that would increase the number of places changes might affect and we have to test. Perhaps a duplicate of `ProductImageThumbnail` for POS would make sense? So the `CartItemRowViewModel` can decide if the image that we load is from cache or not, rather than having this logic in the view.

Thoughts? Happy to iterate here, or in a different PR

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductImageThumbnail.swift:
```diff
@@ -7,27 +7,39 @@ struct ProductImageThumbnail: View {
-                      height: productImageSize * scale),
-                               mode: .aspectFill
-        )
+        if usesDefaultCache {
+            DefaultImageProcessor()
+        } else {
+            ResizingImageProcessor(referenceSize:
+                    .init(width: productImageSize * scale,
+                          height: productImageSize * scale),
+                                   mode: .aspectFill)
+        }
```
The current usage of `ResizingImageProcessor` breaks KF's built-in cache, if we use this processor it only seems to cache very specific images, but most of them are not cached. This is resolved as soon as we pass the `DefaultImageProcessor`. 

I've also confirmed that by removing the processor, other places where the `ProductImageThumbnail` is used (like during product selection on Order creation) images also start to be cached automatically, which currently doesn't happen, making for a better experience. Since this is outside of the scope of the PR and would add many testing cases related to the app and not the POS I've left it outside for the moment (this was added initially to tackle stores with multiple and large images), so we keep using the existing implementation in the app, but the default processor in POS.

staskus commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductImageThumbnail.swift:
```diff
@@ -7,27 +7,39 @@ struct ProductImageThumbnail: View {
     var body: some View {
-        KFImage.url(productImageURL)
+        KFImage
+            .url(productImageURL)
```
```suggestion
            .url(productImageURL)
            .cacheOriginalImage()
```

What would you think of using `cacheOriginalImage`, would it achieve the same result?

We could create a conditional flag from `ProductImageThumbnail` which we could pass from itemRow and itemCard, which would conditionally set `cacheOriginalImage()` modifier:

> If set and an `ImageProcessor` is used, Kingfisher will try to cache both the final result and original
    /// image. Kingfisher will have a chance to use the original image when another processor is applied to the same
    /// resource, instead of downloading it again

It should ensure that not only resized image we use for product list is cached but also original image, so it could be used for resized version in the cart list.

iamgabrielma replied:
Thanks for the recommendation, that's much cleaner, and reduces the PR considerably 💯 

I made the changes on https://github.com/woocommerce/woocommerce-ios/pull/13921/commits/e60bf6e7b774ae7b992d7e00486c7bca0b184f2a , then reverted the logic needed for the initial implementation in the following commits, with this, the `ProductImageThumbnail` continues using cache for the POS views, but keeps the existing behaviour for other uses (using it everywhere looks like a good quick project for hack week, perhaps adding as well some CacheConfiguration file for fine-tuning details).

I wasn't sure of how the combination of `.url()` and `.cacheOriginalImage()` would work together, but it's covered:
> When you set the same URL with KFImage, it will be loaded from the cache [...] When converting to source, a LocalFileImageDataProvider returns the image if the URL points to a local file, or uses the network to download it otherwise.

In order to test, the easier way is to [add an .onSuccess() modifier](https://github.com/woocommerce/woocommerce-ios/blob/bdb7a44f4283ae97f20a9aaaa589862c9dda4ee6/WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductImageThumbnail.swift#L43) to the `KFImage` so the system tells us where the image comes from:

``` diff
.resizable()
+.onSuccess({ result in
+  switch result.cacheType {
+    default:
+      debugPrint("🍍 Image \(productImageURL?.absoluteString) loaded from \(result.cacheType)")
+    }
+})
```

And we get the Image origin both when loaded on Item List and when tapping the items to add to cart. We can also note it uses the original or scaled versions as needed.
```
"🍍 Image Optional(\"https://indiemelon.mystagingwebsite.com/wp-content/uploads/2024/01/IMG_0316.png\") loaded from disk"
"🍍 Image Optional(\"https://indiemelon.mystagingwebsite.com/wp-content/uploads/2023/10/img_0005-scaled.jpeg\") loaded from disk"
"🍍 Image Optional(\"https://indiemelon.mystagingwebsite.com/wp-content/uploads/2023/12/img_0250-scaled.jpeg\") loaded from disk"
```

staskus replied:
Thanks for this explanation and ways of confirming cache behavior!

staskus reviewed (commented):
Thanks for the work! I wonder if we could simplify this, I left a suggestion.

iamgabrielma committed:
Use . cacheOriginalImage modifier on ImageThumbnail

iamgabrielma committed:
Revert changes to load image from cache through VM

iamgabrielma committed:
Clean-up unnecessary import and var declaration

staskus reviewed (approved):
:shipit:
